proxy: check loop with proxyID, endpoint in header

gyuho opened the pull request:
This detects proxy loop by passing around proxyID and endpoints in HTTP header.
Once it detects the loop, it updates proxy's endpoints.

Fixes #3317.

(*updated PR description according the most recent commit)

gyuho commented on proxy/reverse.go:
```diff
@@ -92,6 +94,11 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
 	removeSingleHopHeaders(&proxyreq.Header)
 	maybeSetForwardedFor(proxyreq)
 
+	if err := checkProxyLoop(proxyreq, p.proxyID, 10); err != nil {
```
`10` is just an arbitrary number. I wasn't sure what to put as a limit.

gyuho commented on proxy/reverse.go:
```diff
@@ -92,6 +94,11 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
 
+	if err := checkProxyLoop(proxyreq, p.proxyID, 10); err != nil {
+		plog.Debugf("found proxy loop (%v)", err)
+		return
```
As suggested [here](https://github.com/coreos/etcd/pull/3636#issuecomment-145682590), if it finds a loop, it **returns** without updating endpoints.

yichengq commented:
I think one valuable result of detection is that we could forbid to use the address that points to itself anymore.

gyuho commented:
I see. @yichengq, then where should we start to implement that?

https://github.com/coreos/etcd/pull/3636  also prevents loopback to localhost,
but to detect the recursive behavior of proxying, should we do it in `etcdmain`
and validate the config from flags?

Thanks,

gyuho commented:
I think config validation that I mention should still be in `proxy` package,
rather than `etcdmain` in case `etcd` runs as an API without flags. Which means
the proxy-loop-check should be in `proxy.NewHandler` as https://github.com/coreos/etcd/pull/3636.

Any other suggestion to resolve this issue https://github.com/coreos/etcd/issues/3317?

yichengq commented:
@gyuho I think I don't express my idea clearly, and let me describe it in details:

1. I think it is valuable to detect the proxy loopback when proxying the requests. It could help to stop infinite proxy.
2. If we detect that proxying to url A results in proxy loopback, we could forbid using url A anymore, because it will always result in proxy loopback. This could be a future work.

Do you agree on that?

gyuho commented:
@yichengq yes you are right. I agree and think that's what #3317 is trying to address. 

For future work, it would be helpful if you or Xiang can provide more details on what and how should be done to resolve this issue.

Thanks!

gyuho commented:
Now I see yichen's point. I think if we prevent the proxy loop in the first place (when we configure etcd server) we wouldn't even need this extra http header. I need more research on this. Thanks for feedback!

gyuho commented:
yichen had an idea about updating (*or correcting*) the endpoints when we find the loop from http headers. It can be done within `proxy` package. I will try that in the next coming commits here.

Thanks,

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,39 @@ func maybeSetForwardedFor(req *http.Request) {
+	if v, ok := req.Header["X-Etcd-Proxy-Hop-Seen-By"]; ok {
+		for _, id := range strings.Split(strings.Join(v, ", "), ", ") {
+			if id == proxyID {
+				seen = true
```
If proxy ID has been seen before, could we stop proxying immediately?

gyuho replied:
+1. Ok sure, it sounds good.

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,39 @@ func maybeSetForwardedFor(req *http.Request) {
+		// fold multiple proxyIDs into one header
+		proxyID = strings.Join(v, ", ") + ", " + proxyID
+	}
+	req.Header.Set("X-Etcd-Proxy-Hop-Seen-By", proxyID)
```
Do you have any plan about how to get the endpoint that needs to be forbidden in the follow-up? Could we use as few headers as possible to make it simple?

yichengq replied:
Something could be a reference: https://en.wikipedia.org/wiki/X-Forwarded-For

gyuho replied:
I kept this one header `X-Etcd-Proxy-Hop-Seen-By` to record proxyIDs and detect proxy loop by checking this header in the following requests. And have a separate `struct` member `proxyURLs` to record proxy URLs in general. Once it finds out the proxy loop, it will remove the proxy URLs from the `director`'s endpoints. Thanks,

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,39 @@ func maybeSetForwardedFor(req *http.Request) {
+	req.Header.Set("X-Etcd-Proxy-Hop-Seen-By", proxyID)
+
+	hs := strconv.Itoa(hopLimit)
+	if v, ok := req.Header["X-Etcd-Proxy-Hop-Count"]; ok {
```
Do we really need this header?

gyuho replied:
Thanks for feedback. I just started working on next commit and once we update the endpoints, we wouldn't need this header; instead just stop the proxy.

yichengq replied:
> once we update the endpoints, we wouldn't need this header; instead just stop the proxy.

You mean once we detect it, we would stop proxying requests, right?

i am glad that you could do it one step further. I hope this PR still focuses on detecting proxy loop, and let us finish it step by step.

gyuho commented:
@yichengq Thanks for feedback. I just pushed my second commit.

1. I kept only one http header to handle proxy loops, without hop counting.
So once the proxy id is seen twice, it will be recognized as a proxy loop.
Please let me know if you have any better way to do this.

2. Once the proxy loop is found, it **returns** to exit the method `ServeHTTP`.
Is this enough to stop serving proxy requests?

3. I added a function to get loopback IP to handle `localhost` and `127.0.0.1` or such.
I did this because I wasn't sure how to handle `localhost` being changed to `127.0.0.1`,
when given `localhost` in `-advertise-client-urls`. And `localhost` is handled as a special
case in my code.

4. I added lines to update `endpoints` by calling `refresh` method after it updates the
`banned` map, which is a map of proxy host URL string that needs to be banned, for now
mostly from proxy loop.

You can test the behavior of looping proxy with the following script:

```bash
# in my local machine
export INFRA_IP=127.0.0.1;
etcd \
-name infra0 \
-initial-cluster-token my-infra \
-initial-cluster infra0=http://$INFRA_IP:2380 \
-initial-cluster-state new \
-initial-advertise-peer-urls http://$INFRA_IP:2380 \
-listen-peer-urls http://$INFRA_IP:2380 \
-listen-client-urls http://$INFRA_IP:2379 \
-advertise-client-urls http://localhost:8080 \
;

export INFRA_IP=127.0.0.1;
etcd \
-proxy on \
-initial-cluster infra0=http://$INFRA_IP:2380 \
-listen-client-urls http://localhost:8080 \
;

curl -L http://localhost:8080/v2/keys;
curl -L -XPUT http://localhost:8080/v2/keys/test -d value="Hello";

```

yichengq commented:
The method looks complicated because it needs to detect loopback IP, handle localhost specially, and remote address seen by serveHTTP doesn't always match forward address.

Could we use the header to record the forward/proxy addresses?

gyuho commented:
I agree. I will try to simplify this in the next commit. Thanks!

gyuho commented:
1. I removed the lines for getting loopback IP.
2. Now instead it uses proxy URLs from `proxy.NewHandler` when updating endpoints.
3. It still uses http header of proxy ID to detect a proxy loop.

Please let me know if you have any feedback.
(Sorry for the messy commits. I will squash all commits into one later, once this gets reviewed.)

gyuho commented on proxy/reverse.go:
```diff
@@ -92,6 +103,13 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
+	// same proxyID which was seen before, we stop serving
+	// the proxy request by exiting ServeHTTP method.
+	if p.checkProxyLoop(proxyreq) {
+		return
```
this `return`s from `ServeHTTP` and it would stop serving the proxy requests.

gyuho commented on proxy/proxy.go:
```diff
@@ -42,9 +48,23 @@ type GetProxyURLs func() []string
 // which will proxy requests to an etcd cluster.
 // The handler will periodically update its view of the cluster.
 func NewHandler(t *http.Transport, urlsFunc GetProxyURLs, failureWait time.Duration, refreshInterval time.Duration) http.Handler {
+	proxyURLs := make(map[string]struct{})
```
I use the type `map[string]struct{}` to make it same as the one in `director` so that the director can use its `banned` which is also `map[string]struct{}` to remove cyclic proxy URLs from its endpoints (when calling `refresh` or `endpoints` method).

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +221,37 @@ func maybeSetForwardedFor(req *http.Request) {
+	// if there is a proxy loop, update endpoints.
+	if seen {
+		p.Lock()
+		p.director.banned = p.proxyURLs
```
This is totally wrong.

When ServeHTTP, it asks director to get current endpoint list, and tries to redirect to them one by one. The endpoint list in director is refreshed every some interval to make it sync with the advertised client urls in the cluster, which consists of advertised client urls of all members. When you ban proxy loop, you should only ban the one that leads to the loop.

My thought is to record the url used to proxy in `X-Etcd-Proxy-Hop-Seen-By`, so we could know which endpoint should be banned.

gyuho replied:
This makes sense. Thanks for clarification. I will try to address this in the next commit.

Thanks!

gyuho commented:
The last commit is **not to be merged**. Only for code review. I will squash all previous commits into one later, once this gets reviewed.

Explanation coming.

gyuho commented:
/cc @yichengq  Sorry for delay (I am currently in a different timezone.)

I pushed another commit to `record the url used to proxy in X-Etcd-Proxy-Hop-Seen-By`
as you suggest. It works to detect the proxy loop but I cannot think of an easy way to
get the actual proxy endpoint strings, such as `http://localhost:8080` from the URLs that
it detects as a loop. It detects from `request.RemoteAddr` which is resolved from `localhost`.
Without loopback function in my previous commits, there is no way that we can find the
actual resolved address and map it back to `localhost` endpoint. I keep the map `banned`
later to use this to update endpoints.

I might be still misunderstanding your point. If so, could you please clarify more on your saying
`record the url used to proxy in X-Etcd-Proxy-Hop-Seen-By`? What would it be different than
`X-Forwarded-For`? Do we also record the port?

Thanks,

<br>
<hr>
For reference, this is the sample output from current implementation.

```
addr, clientIP: 127.0.0.1:36843 127.0.0.1
addr, clientIP: 127.0.0.1:36844 127.0.0.1

banned: map[127.0.0.1:36844:{}]

proxyreq.Header: map[User-Agent:[curl/7.35.0] 
Accept:[*/*]
X-Etcd-Proxy-Hop-Seen-By:[127.0.0.1, 127.0.0.1]
X-Forwarded-For:[127.0.0.1, 127.0.0.1]
Accept-Encoding:[gzip]]

endpoint: http://localhost:8080

```

gyuho commented on proxy/director.go:
```diff
@@ -58,6 +64,10 @@ func (d *director) refresh() {
 		}
+		host, port, _ := net.SplitHostPort(uu.Host)
+		if _, ok := d.banned[host+":"+port]; ok {
+			continue
```
when used with `port` and `proxy`, this is currently useless.
I am waiting on code review for better ways to update `endpoint`.

gyuho commented on proxy/reverse.go:
```diff
@@ -92,6 +92,23 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
+		fmt.Println("banned:", p.director.banned)
+		fmt.Printf("proxyreq.Header: %+v\n", proxyreq.Header)
+		for _, v := range p.director.ep {
+			fmt.Printf("endpoint: %s\n", v.URL.String())
```
print statements for debugging... to be removed later.

yichengq commented:
@gyuho My draft thought is to put the proxyID and url (that proxy used to forward) to some header. When it finds that proxyID has appeared, it banned the url with the ID.

For example, proxy A with id 1234 forwards request to 127.0.0.1:4001, so it has header with value `(1234, 127.0.0.1:4001)`. Then proxy A receives this request again and find that itself(1234) proxied it before. So proxy A drops it and banned endpoint `127.0.0.1:4001`.

It doesn't use RemoteAddr at all, because RemoteAddr is not always equal to endpoint used by proxy.

gyuho commented:
Thanks for clarification! I will try that in the next commit.

gyuho commented on proxy/reverse.go:
```diff
@@ -92,6 +96,16 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
+		// update endpoints by calling refresh method
+		p.director.refresh()
+		plog.Debugf("found proxy loop (%v)", seen)
+	}
```
Should we just return before we get to this point?

gyuho commented on proxy/reverse.go:
```diff
@@ -60,6 +60,10 @@ func removeSingleHopHeaders(hdrs *http.Header) {
+
+	// proxyID maps to the endpoints of the proxy.
+	proxyID   string
+	endpoints map[string]struct{}
```
I don't we need to pass the endpoints to http header either. We can only pass the `proxyID` and detect by ID, and once we detect it, we map to `endpoints` from this struct and ban those from `p.director.endpoints` with `refresh` method.

gyuho commented:
/cc @yichengq 

I think we don't need to pass the **urls** to **HTTP header**. We can only pass the `proxyID` and detect by ID, and once we detect it, we map to `endpoints` from `reverseProxy` struct and ban those from `p.director.endpoints` with `refresh` method. Does this make sense?

To sum, the goal is to detect proxy and update its endpoints so that we don't get into cyclic proxy.

1. It detects proxy when it sees a proxyID in the HTTP header twice.
2. The proxy forwards its request to the endpoints in `p.director`, which is to be updated by `refresh` method. So we need a way to tell which endpoints to ban. So I defined `bannedEndpoints` member in the `director` and everytime `refresh` method gets called, it removes the endpoints from `bannedEndpoints` and initialize this member so that it gets called only once to ban the proxy loop.

I am about to board a plane, but please let me know if I misunderstood again :(

gyuho commented on proxy/reverse.go:
```diff
@@ -92,6 +96,16 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
+
+		// update endpoints by calling refresh method
+		p.director.refresh()
+		plog.Debugf("found proxy loop (%v)", seen)
```
Should we just return before we get to this point? `refresh` call in line 105 removes the looping endpoints so it shouldn't matter though.

gyuho commented:
I thought about it and I think this is also against your previous comment:

> This is totally wrong.
>
> When ServeHTTP, it asks director to get current endpoint list, and tries to redirect to them one by one. The endpoint list in director is refreshed every some interval to make it sync with the advertised client urls in the cluster, which consists of advertised client urls of all members. When you ban proxy loop, you should only ban the one that leads to the loop.
>
> My thought is to record the url used to proxy in X-Etcd-Proxy-Hop-Seen-By, so we could know which endpoint should be banned.

I think I am failing to `ban proxy loop, you should only ban the one that leads to loop`. I will try to fix this.

gyuho commented:
/cc @yichengq 

As you suggested, I am passing both proxyID and URL to HTTP headers.
It got more verbose though. Please let me know if you have any feedback.

gyuho commented:
Pushed some updates.

1. squash commits into one
2. `checkProxyLoop` in method (not in function)
3. change its method description

Thanks!

gyuho commented:
Will fix. Current code seems to remove all endpoints returning `proxy: zero endpoints currently available` from this one loop.

yichengq commented on proxy/reverse.go:
```diff
@@ -134,6 +135,11 @@ func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request
 		}
 		redirectRequest(proxyreq, ep.URL)
 
+		if repeated := p.checkProxyLoop(proxyreq, ep.URL.String()); repeated {
```
could we move this check to the beginning of `ServeHTTP`? If it cannot pass the check, there is no need to try to proxy this request. It makes the logic more straightforward.

gyuho replied:
@yichengq As far as I understand, when given multiple endpoints, we proxy a request only to one of the endpoints as in:

```go
for _, ep := range endpoints {
   ...
   res, err = p.transport.RoundTrip(proxyreq)
   ...
   break
}
```
So looks like this is where the looping is happening, and best place to detect if an endpoint
gets cycled with that transport. If we put this check to the beginning of the `ServeHTTP`, we
won't be able to know which endpoint gets the RoundTrip.

Or I might be wrong. Please correct me. Thanks!

yichengq replied:
If local proxyID appears in the header of incoming http request, this indicates that something wrong has happened, and the associated endpoint should be banned. So there is no need to wait until proxying phase to detect it IMO.

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +209,33 @@ func maybeSetForwardedFor(req *http.Request) {
+// and endpoints, and call refresh methods to update endpoints
+// if there's any proxy loop found. It returns true if it has
+// found a proxy loop from a endpoint string.
+func (p *reverseProxy) checkProxyLoop(req *http.Request, ep string) bool {
```
This name does not match its behavior. The function does check, set and ban. Either you should find an good naming about it, or you should split it into multiple functions.

gyuho replied:
+1 for splitting into multiple functions. I will change that. And also I was just changing the argument type just to `url.URL` to make it consistent with `redirectRequest`. Thanks

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +209,33 @@ func maybeSetForwardedFor(req *http.Request) {
+// found a proxy loop from a endpoint string.
+func (p *reverseProxy) checkProxyLoop(req *http.Request, ep string) bool {
+	// concatenation of proxyID and endpoint URL string.
+	pv := p.proxyID + "___" + ep
```
Is "___" a good enough concatenation way in http header value?

gyuho replied:
Actually, `___` looks too random. I wasn't sure how to pass these two ids to headers without introducing extra headers. Any suggestion?

yichengq replied:
not really. could you investigate the way that other headers handles this? It could be a good reference for us.

gyuho replied:
Sure, I will do more research and come back with another commit. Thanks for the review!

gyuho commented:
/cc @yichengq

I updated `___` part with JSON, if it's ok. I couldn't find any example that
does this, because standard way is to have only one key (multi-values) in one
header. It might be better to keep the previous hacky way.

And I have one more question:

For the logic, `curl -L http://localhost:8080/v2/keys;` request would only contains
`/v2/keys` at the top of `ServeHTTP` (no Scheme or Host of  `endpoint` is known at
this point). And if the goal of this loop detection is to remove the looping `endpoint`,
we must start detection from `endpoint` slice. As Yichen pointed out, the `RemoteAddr`
would not work when the loop happens in localhost.

Then I am not sure how we could safely remove only the cycling endpoint if we detect
it by proxyID. We also need `endpoint`. And the only way to know which endpoint is
being forwarded to is place `checkProxyLoop` after `redirectRequest(proxyreq, ep.URL)`,
I think.

Please correct me if I am wrong.

Thanks,

yichengq commented:
> I updated ___ part with JSON, if it's ok

I would say even `%ID%=%URL%` is better than JSON and more readable.

> Then I am not sure how we could safely remove only the cycling endpoint if we detect
it by proxyID.

We record (proxyID, endpoint) pair in the header value. Because proxyID is unique, when we find that proxyID matches local ID, we could remove the endpoint in the pair. Does this make sense?

gyuho commented:
@yichengq Got it! Thanks for review. I will try that in the next commit.

gyuho commented:
@yichengq Sorry for delaying.

1. I moved the proxy check method at the beginning of the `ServeHTTP` method,
as you pointed out. I was wrong.

2. I had to add `recordProxy` function to keep track of endpoints. Is that ok?

I really want to get this done. Please ping me anytime if you have any feedback.

Thanks!

gyuho commented:
kindly ping @yichengq 

Please let me know if you still have some feedback. Thanks yichen!

yichengq commented on proxy/reverse.go:
```diff
@@ -60,9 +60,15 @@ func removeSingleHopHeaders(hdrs *http.Header) {
 
 func (p *reverseProxy) ServeHTTP(rw http.ResponseWriter, clientreq *http.Request) {
+	if rp, repeated := p.checkProxyLoop(clientreq); repeated {
+		plog.Debugf("found proxy loop (%s)", rp)
```
maybe `Warningf` because this is critical.

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,42 @@ func maybeSetForwardedFor(req *http.Request) {
 	req.Header.Set("X-Forwarded-For", clientIP)
 }
+
+const proxyLoopHeaderKey = "X-Etcd-Proxy-Hop-Seen-By"
```
move this just below the import?

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,42 @@ func maybeSetForwardedFor(req *http.Request) {
+func recordProxy(req *http.Request, proxyID string, loc url.URL) {
+	pv := proxyID + "=" + loc.String()
+	if prior, ok := req.Header[proxyLoopHeaderKey]; ok {
+		pv = strings.Join(prior, ", ") + ", " + pv
```
i would like `strings.Join(append(prior, pv), ", ")` because you don't need to write `", "` twice.

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,42 @@ func maybeSetForwardedFor(req *http.Request) {
+
+// checkProxyLoop detects a proxy loop by passing around proxyID
+// and endpoints. It returns the repeating endpoint string and
+// truu if it finds the loop.
```
true

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,42 @@ func maybeSetForwardedFor(req *http.Request) {
+// checkProxyLoop detects a proxy loop by passing around proxyID
+// and endpoints. It returns the repeating endpoint string and
+// truu if it finds the loop.
+func (p *reverseProxy) checkProxyLoop(req *http.Request) (string, bool) {
```
could we make this a helper function instead of a member function?

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,42 @@ func maybeSetForwardedFor(req *http.Request) {
+				if p.proxyID == vs[0] {
+					return vs[1], true
+				}
+			}
```
what may happen if `len(vs) != 2`?

gyuho replied:
It will be ignored, because we don't want to ban either empty endpoint or unknown format of proxy loop header. Is that ok?

yichengq replied:
If it has unknown format, it looks like weird IMO.

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +210,42 @@ func maybeSetForwardedFor(req *http.Request) {
+	p.director.Unlock()
+
+	// update endpoints by calling refresh method.
+	p.director.refresh()
```
why do we refresh here?

gyuho replied:
Thanks! This updates `bannedEndpoints` and calls `refresh`, which checks if the endpoint is found in `bannedEndpoints`. And the refresh method excludes the ones that are in `bannedEndpoints`.

To sum, we call refresh to exclude cyclic endpoint.

yichengq replied:
i see it has been excluded when calling `endpoints()` function, right?

gyuho replied:
@yichengq Yeah it excludes both from refresh and endpoints to make sure.

https://github.com/gyuho/etcd/blob/check_proxy_loop/proxy/director.go#L101-L103


What do you think?

yichengq replied:
I think there is no need to refresh it here, because ban is good enough. moreover, i think this should be a member function of `director`

gyuho replied:
Good idea. I will try that. Thanks,

yichengq commented:
The general approach is good! I think it is almost there.

gyuho commented:
@yichengq Your feedback has been addressed.
Pleas take a look again. Thanks for feedback!

gyuho commented on proxy/reverse.go:
```diff
@@ -203,3 +212,41 @@ func maybeSetForwardedFor(req *http.Request) {
+
+	// TODO: immediately update endpoints by
+	// calling refresh method.
+	// d.refresh()
```
@yichengq I commented this lines out. What do you think?
I agree but if we don't exclude the endpoint here by calling refresh,
we would still have proxy loop until the next refresh call, which is 
by default 30-second.

yichengq replied:
I see you exclude banned endpoints in director.endpoints() https://github.com/coreos/etcd/pull/3645/files#diff-eb83fa9ddcf3dba0b49cf06cddd60dc1R96, and it is the function to get endpoint list when proxying request at https://github.com/coreos/etcd/pull/3645/files#diff-11170a80f9a71a746a713b993f18e4b3R103 . So i think proxy after ban should not trigger proxy loop.

gyuho replied:
You are right. Thanks! I will delete those comments.

gyuho replied:
Just fixed.

gyuho commented on proxy/reverse.go:
```diff
@@ -203,3 +212,41 @@ func maybeSetForwardedFor(req *http.Request) {
+}
+
+// banEndpoint bans the URL from reverseProxy's endpoints.
+func (d *director) banEndpoint(rp string) {
```
And I changed it to `director` method.

yichengq replied:
please move it to director struct in director.go~

gyuho commented:
@yichengq I moved `banEndpoint` method to `director.go`.
Please let me know if you have any feedback! Thanks!

yichengq commented on proxy/reverse.go:
```diff
@@ -203,3 +212,30 @@ func maybeSetForwardedFor(req *http.Request) {
+				if proxyID == vs[0] {
+					return vs[1], true
+				}
+			}
```
could we add a log line if `len(vs) != 2`?

yichengq commented:
lgtm. defer to @xiang90 

Have you manually tested it?

gyuho commented:
@yichengq I changed it to print warning message and I tested it with the following commands:

```sh
# in my local machine
export INFRA_IP=127.0.0.1;
etcd \
-name infra0 \
-initial-cluster-token my-infra \
-initial-cluster infra0=http://$INFRA_IP:2380 \
-initial-cluster-state new \
-initial-advertise-peer-urls http://$INFRA_IP:2380 \
-listen-peer-urls http://$INFRA_IP:2380 \
-listen-client-urls http://$INFRA_IP:2379 \
-advertise-client-urls http://localhost:8080 \
;

export INFRA_IP=127.0.0.1;
etcd \
-proxy on \
-initial-cluster infra0=http://$INFRA_IP:2380 \
-listen-client-urls http://localhost:8080 \
;

curl -L http://localhost:8080/v2/keys;
curl -L -XPUT http://localhost:8080/v2/keys/test -d value="Hello";
```

And it logs like:

```
2015-11-24 16:40:22.691937 W | proxy: found proxy loop (http://localhost:8080)
2015-11-24 16:40:28.800697 I | proxy: zero endpoints currently available
```
